File: correct-crypto/javascript/passwordHash.js
// Correct password hashing with scrypt and a random salt (Node.js).
'use strict';
const crypto = require('crypto');

const PARAMS = { N: 2 ** 14, r: 8, p: 1 };

function hashPassword(password) {
  const salt = crypto.randomBytes(16);
  const dk = crypto.scryptSync(password, salt, 32, PARAMS);
  return Buffer.concat([salt, dk]);
}

function verifyPassword(stored, password) {
  const salt = stored.subarray(0, 16);
  const expected = stored.subarray(16);
  const dk = crypto.scryptSync(password, salt, 32, PARAMS);
  return crypto.timingSafeEqual(dk, expected);          // constant-time
}

const record = hashPassword('correct horse');
console.log('scrypt verify OK:', verifyPassword(record, 'correct horse'));
module.exports = { hashPassword, verifyPassword };

File: correct-crypto/ruby/password_hash.rb
# frozen_string_literal: true
# Correct password hashing with scrypt and a random salt (OpenSSL::KDF).
require 'openssl'
require 'securerandom'

def hash_password(password)
  salt = SecureRandom.random_bytes(16)
  dk = OpenSSL::KDF.scrypt(password, salt: salt, N: 2**14, r: 8, p: 1, length: 32)
  salt + dk
end

def verify_password(stored, password)
  salt = stored[0, 16]
  expected = stored[16..]
  dk = OpenSSL::KDF.scrypt(password, salt: salt, N: 2**14, r: 8, p: 1, length: 32)
  OpenSSL.fixed_length_secure_compare(dk, expected) # constant-time
end
